openwisp_radius/tests/test_commands.py

```diff
@@ -140,14 +140,21 @@ def test_prefix_add_users_commnad(self):
         users = get_user_model().objects.all()
         self.assertEqual(users.count(), 10)
         for u in users:
-            self.assertTrue('openwisp' in u.username)
+            self.assertTrue('test-prefix7' in u.username)
         self.assertEqual(radiusbatch.expiration_date.strftime('%d-%m-%y'), '28-01-18')
-        options = dict(organization=self.default_org, prefix='gsoc', n=5, name='test1')
+        options = dict(
+            organization=self.default_org, prefix='test-prefix8', n=5, name='test1'
+        )
         self._call_command('prefix_add_users', **options)
         self.assertEqual(RadiusBatch.objects.all().count(), 2)
         self.assertEqual(get_user_model().objects.all().count(), 15)
-        with self.assertRaises(SystemExit):
-            options = dict(
-                organization=self.default_org, prefix='openwisp', n=-5, name='test2'
-            )
+        self._delete_file('test-prefix8.pdf')
+        options = dict(
+            organization=self.default_org, prefix='test-prefix7', n=-5, name='test2'
+        )
+        try:
             self._call_command('prefix_add_users', **options)
+        except SystemExit:
+            self._delete_file('test-prefix7.pdf')
+        else:
+            self.fail('SystemError not raised!')
```

Comment:
Even when _call_command fails and the radiusbatch is not created, the file is created. Is this intended behavior?

Reply:
I don't think so, looks like a bug to me.

Reply:
Look here: https://github.com/openwisp/openwisp-radius/blob/master/openwisp_radius/base/models.py#L895-L905

`generate_pdf` writes to disk before validation, so if the validation fails, the file is left there anyway. Can you help to fix this please? We could do something like:

```python
try:
    batch.full_clean()
except ValidationError:
    delete_file()  # (pseudo code)
    raise e
```
I also created an issue for the GCI to address the broader issue that we should not save this info in a file on disk, I thought about a way to avoid this that should work: #102. 
Shame it will be a backward incompatible change, but I don't think it's a huge issue.

Reply:
@nemesisdesign I've taken a different approach to solve it than in the comment:

In [`generate_pdf` I changed the path to `/tmp/{prefix}.pdf`](https://github.com/openwisp/openwisp-radius/pull/100/files#diff-d4dfe472aa734b022012d64055559459R162 ).

So that the file created before save goes to `/tmp/` and os cleans it later while the [final file saved by django goes in `{settings.MEDIA_ROOT}/{prefix}.pdf`](https://github.com/openwisp/openwisp-radius/pull/100/files#diff-141584df2cf4ec556f1b9fe45e78d246R902)

This also takes care of the problem that 2 files are create: one by `generate_pdf()` and another by django with a random name which is saved with the model.
Now, the name of the pdf file and radiusbatch name can be the same.

What do you think about this approach?

Reply:
P.S: I'll work on #103 as soon as possible! :smile:

Reply:
> This also takes care of the problem that 2 files are create: one by `generate_pdf()` and another by django with a random name which is saved with the model.
> Now, the name of the pdf file and radiusbatch name can be the same.
> 
> What do you think about this approach?

Is good thanks!

> P.S: I'll work on #103 as soon as possible!

Did you mistype? #103 is related to openwisp-notifications.

Reply:
Yeah, I meant #102